src/lib/pgp-key.cpp

```diff
@@ -961,6 +969,373 @@ pgp_key_get_subsig(const pgp_key_t *key, size_t idx)
+            pgp_signature_info_t sinfo = {};
+            sinfo.sig = &sig->sig;
+            sinfo.signer = key;
+            sinfo.signer_valid = true;
+            pgp_userid_t *uid = pgp_key_get_userid(key, sig->uid);
+            if (uid) {
+                signature_check_certification(&sinfo, pgp_key_get_pkt(key), &uid->pkt);
+                sig->validated = true;
+                sig->valid = sinfo.valid && !sinfo.expired;
+            }
+            continue;
+        }
+
+        if (pgp_sig_is_key_revocation(key, sig)) {
+            pgp_signature_info_t sinfo = {};
+            sinfo.sig = &sig->sig;
+            sinfo.signer = key;
+            sinfo.signer_valid = true;
+            signature_check_direct(&sinfo, pgp_key_get_pkt(key));
+            sig->validated = true;
+            sig->valid = sinfo.valid;
+            /* revocation signature cannot expire */
+        }
+    }
+}
+
+static void
+pgp_subkey_validate_self_signatures(pgp_key_t *sub, pgp_key_t *key)
+{
+    for (size_t i = 0; i < pgp_key_get_subsig_count(sub); i++) {
+        pgp_subsig_t *sig = pgp_key_get_subsig(sub, i);
+        if (sig->validated) {
+            continue;
+        }
+
+        if (pgp_sig_is_subkey_binding(sub, sig)) {
+            pgp_signature_info_t sinfo = {};
+            sinfo.sig = &sig->sig;
+            sinfo.signer = key;
+            sinfo.signer_valid = true;
+            signature_check_binding(&sinfo, pgp_key_get_pkt(key), pgp_key_get_pkt(sub));
+            sig->validated = true;
+            sig->valid = sinfo.valid && !sinfo.expired;
+            continue;
+        }
+
+        if (pgp_sig_is_subkey_revocation(sub, sig)) {
+            pgp_signature_info_t sinfo = {};
+            sinfo.sig = &sig->sig;
+            sinfo.signer = key;
+            sinfo.signer_valid = true;
+            signature_check_subkey_revocation(
+              &sinfo, pgp_key_get_pkt(key), pgp_key_get_pkt(sub));
+            sig->validated = true;
+            sig->valid = sinfo.valid;
+            /* revocation signature cannot expire */
+        }
+    }
+}
+
+static pgp_map_t ss_rr_code_map[] = {
+  {PGP_REVOCATION_NO_REASON, "No reason specified"},
+  {PGP_REVOCATION_SUPERSEDED, "Key is superseded"},
+  {PGP_REVOCATION_COMPROMISED, "Key material has been compromised"},
+  {PGP_REVOCATION_RETIRED, "Key is retired and no longer used"},
+  {PGP_REVOCATION_NO_LONGER_VALID, "User ID information is no longer valid"},
+  {0x00, NULL}, /* this is the end-of-array marker */
+};
+
+bool
+pgp_subkey_refresh_data(pgp_key_t *sub, pgp_key_t *key)
+{
+    /* validate self-signatures if not done yet */
+    if (key) {
+        pgp_subkey_validate_self_signatures(sub, key);
+    }
+    pgp_subsig_t *sig = pgp_key_latest_binding(sub, key);
+    /* subkey expiration */
+    sub->expiration = sig ? signature_get_key_expiration(&sig->sig) : 0;
+    /* subkey flags */
+    if (sig && signature_has_key_flags(&sig->sig)) {
+        sub->key_flags = sig->key_flags;
```

Comment:
It seems like this would leave an incorrect value here in the case where a `PGP_SIG_SUBPKT_KEY_FLAGS` subpacket was previously present and has now been removed, wouldn't it?

Reply:
Agree, thanks for spotting this. Fixed.